_posts/2025-04-25-secure-mcp-sse-server.adoc

```diff
@@ -0,0 +1,399 @@
+
+[source,xml]
+----
+<dependency>
+    <groupId>io.quarkiverse.mcp</groupId>
+    <artifactId>quarkus-mcp-server-sse</artifactId> <1>
+    <version>1.1.1</version>
+</dependency>
+
+<dependency>
+    <groupId>io.quarkus</groupId>
+    <artifactId>quarkus-oidc</artifactId> <2>
+</dependency>
+----
+<1> `quarkus-mcp-server-sse` is required to support MCP SSE transport.
+<2> `quarkus-oidc` is required to secure access to MCP SSE endpoints.
+
+[[tool]]
+=== Tool
+
+Let's create a tool that can be invoked only if the current MCP request is authenticated:
+
+[source,java]
+----
+package org.acme;
+
+import io.quarkiverse.mcp.server.TextContent;
+import io.quarkiverse.mcp.server.Tool;
+import io.quarkus.security.Authenticated;
+import io.quarkus.security.identity.SecurityIdentity;
+import jakarta.inject.Inject;
+
+public class ServerFeatures {
+
+    @Inject
+    SecurityIdentity identity;
+
+    @Tool(name = "user-name-provider", description = "Provides a name of the current user") <1>
+    @Authenticated <2>
+    TextContent provideUserName() {
+        return new TextContent(identity.getPrincipal().getName()); <3>
+    }
+}
+----
+<1> Provide a tool that can return a name of the current user. Note the `user-name-provider` tool name, you will use it later for a tool call.
+<2> Require authenticated tool access. See also how the main MCP SSE endpoint is secured in the <<initial-configuration>> section below.
+<3> Use the injected `SecurityIdentity` to return the current user's name.
+
+[[initial-configuration]]
+=== Configuration
+
+Finally, let's configure our secure MCP server:
+
+[source,properties]
+----
+quarkus.http.auth.permission.authenticated.paths=/mcp/sse <1>
+quarkus.http.auth.permission.authenticated.policy=authenticated
+----
+<1> Enforce an authenticated access to the main MCP SSE endpoint during the initial handshake. See also how the tool is secured with an annotation in the <<tool>> section above, though you can also secure access to the tool by listing both main and tools endpoints in the configuration, for example: `quarkus.http.auth.permission.authenticated.paths=/mcp/sse,/mcp/messages/*`.
+
+We are ready to test our secure MCP server in the dev mode.
+
+== Step 2 - Access MCP server in the dev mode
+
+=== Start MCP server in the dev mode
+
+[source,shell]
+----
+mvn quarkus:dev
+----
+
+The configuration properties that we set in the <<initial-configuration>> section above are sufficient to start the application in the dev mode.
+
+The OIDC configuration is provided in the dev mode automatically by https://quarkus.io/guides/security-openid-connect-dev-services[Dev Services for Keycloak]. It creates a default realm, client and adds two users, `alice` and `bob`, for you to get started with OIDC immediately. You can also register a custom Keycloak realm to work with the existing realm, client and user registrations.
+
+No problems if you do not work with Keycloak, see the <<mcp-server-devui>> section for more details.
```

Comment:
i'm not folowing what that sentence is trying to say? If we are running in devmode why is "working with keycloak"  a problem?

Reply:
@maxandersen I'm trying to say that Auth0 users, GitHub users can also use this DevUI and login to their providers from Dev UI, but I'll try to rephase it

Reply:
@maxandersen 
The reason I don't show GitHub in devmode is that for that to work, users would have to allow a redirect to OIDC Dev UI, something like `http://localhost:8080/q/dev-ui/io.quarkus.quarkus-oidc/github-provider` (or similar), see https://quarkus.io/guides/security-openid-connect-dev-services#dev-ui-all-oidc-providers (it shows Google) but also in our Auth0 guide (the link is in the section below here).

This can be done, but with GitHub, they only allow a single redirect URL, but we also need to support a Login endpoint redirect in prod, so it can become messy, users having to update their GitHub OAuth2 registrations dependeing on whether it is devmode or prod....

Reply:
So yes, in Dev UI, we have all out of the box already, and showing Keycloak is simpler for the initial check